GT/GP/dosage entry field options for importing BGEN v1.2

jbloom22 opened the pull request:
The default options are all false. If left all false, the user gets an clear error saying that they must include at least one entry field. This forces users to think about what they actually need to import, as it can make a big difference on, say, UKBB until we have better tech. I've updated the docs and tests accordingly.

@cseed suggested that we remove BGEN v1.1 support if nobody is reliant on it anymore. I've asked on Slack. So I didn't add more complexity to support these options for BGEN v1.1. Rather this PR requires GT and GP set to true if any file is 1.1 (as explained in docs and error message). If nobody minds, we can rip out BGEN 1.1 and update the docs simultaneously in a subsequent PR.

jbloom22 commented on src/main/scala/is/hail/io/bgen/BgenRecord.scala:
```diff
@@ -142,6 +143,8 @@ class BgenRecordV12(compressed: Boolean, nSamples: Int, tolerance: Double) exten
     assert(nBitsPerProb >= 1 && nBitsPerProb <= 32, s"Value for nBits must be between 1 and 32 inclusive. Found $nBitsPerProb.")
 
     val nGenotypes = triangle(nAlleles)
+    val hetRefIndices = (0 until nGenotypes).filter(i => Call.isHetRef(Call2.fromUnphasedDiploidGtIndex(i))).toArray
```
You're totally right, the projection is unnatural.

tpoterba commented on python/hail/methods/impex.py:
```diff
@@ -548,27 +548,50 @@ def grep(regex, path, max_count=100):
            reference_genome=nullable(GenomeReference),
            contig_recoding=nullable(dictof(strlike, strlike)))
-def import_bgen(path, tolerance=0.2, sample_file=None, min_partitions=None, reference_genome=None,
+def import_bgen(path, tolerance=0.2, sample_file=None, gt_field=False, gp_field=False, dosage_field=False,
```
was going to say the same, I think this is a much nicer (and more extensible) interface

jbloom22 replied:
I agree, making the change.

jbloom22 commented:
I've addressed the two comments: now using an `entry_fields` parameter and throwing an error if 'dosage' is requested and any variant is multi-allelic. Docs updated accordingly.

I considered setting dosage on multi-allelics to missing rather than throwing an error, but I think error is safest since I could imagine the missingness leading to QC confusion, and if users want dosage on multi-allelics than they should either use a custom expression or split and then use `hl.gp_dosage`.

jigold commented on src/test/scala/is/hail/io/LoadBgenSuite.scala:
```diff
@@ -261,16 +262,42 @@ class LoadBgenSuite extends SparkSuite {
+      val nNull = dosages.count(_ == null)
+      val nTotal = dosages.length
+      
+      assert(nNull > 0 && nNull < nTotal)
```
Can you hardcode the number of null dosages you expect? All of the v1.2 bgen files (*.8bits.*, *.10bits.*) are the same data as example.v11.bgen and example.v11.gen.

jbloom22 replied:
The number is 2 is the v1.2 files and 1 in the v1 file. I added the v11 file and simplified the last few lines of the test to emphasize that I really just care that there is at least one null value (so it's tested) and that null values occur in the same samples via both routes to dosage.

jbloom22 commented on src/main/scala/is/hail/HailContext.scala:
```diff
@@ -251,15 +251,21 @@ class HailContext private(val sc: SparkContext,
   def importBgen(file: String,
     sampleFile: Option[String] = None,
     tolerance: Double = 0.2,
+    gtField: Boolean,
```
Yes, that's much better.

jbloom22 commented on python/hail/methods/impex.py:
```diff
@@ -547,27 +548,41 @@ def grep(regex, path, max_count=100):
 @typecheck(path=oneof(strlike, listof(strlike)),
            tolerance=numeric,
            sample_file=nullable(strlike),
+           entry_fields=listof(strlike),
```
If it's okay with you, I'll leave it, since the error messages won't be as clear and I don't want to soup up the typecheckers in this PR.

jbloom22 commented:
@jigold addressed your comments and, after seeing how simple it'd be to do, I came around to your point that I might as well make the same change in v1.1. Simplified the docs accordingly. We still may want to drop v1.1 at some point but Cotton agrees there is no urgency while it's not causing problems.

jigold commented on src/test/scala/is/hail/io/LoadBgenSuite.scala:
```diff
@@ -278,25 +278,22 @@ class LoadBgenSuite extends SparkSuite {
-      assert(nNull == dosagesFromGP.count(_ == null))
-      assert(dosages.length == dosagesFromGP.length)      
+      assert(dosages.length == 500 && dosagesFromGP.length == 500)
+      assert(dosages.count(_ == null) > 0)
```
Still not sure why you don't know what the number of `null` calls should be.
